hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -402,35 +438,113 @@ public void addConnectionHeartbeatListener(ConnectionHeartbeatListener connectio
+                        failed(target, connection, e);
+                        callback.onFailure(e);
+                        break;
+                    default:
+                        AuthenticationException exception =
+                                new AuthenticationException("Authentication status code not supported. status:"
+                                        + authenticationStatus);
+                        failed(target, connection, exception);
+                        callback.onFailure(exception);
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                failed(target, connection, t);
+                callback.onFailure(t);
+            }
+        }, executionService.getInternalExecutor());
+    }
+
     private class InitConnectionTask implements Runnable {
 
         private final Address target;
-        private final Address remoteAddress;
-        private final Authenticator authenticator;
+        private final boolean asOwner;
+        private final AuthenticationCallback callback;
 
-        InitConnectionTask(Address target, Address remoteAddress, Authenticator authenticator) {
+        InitConnectionTask(Address target, boolean asOwner, AuthenticationCallback callback) {
             this.target = target;
-            this.remoteAddress = remoteAddress;
-            this.authenticator = authenticator;
+            this.asOwner = asOwner;
+            this.callback = callback;
         }
 
         @Override
         public void run() {
-            final Object lock = getLock(target);
-            synchronized (lock) {
-                ClientConnection connection = connections.get(target);
-                if (connection != null) {
-                    return;
-                }
+            ClientConnection connection = connections.get(target);
+            if (connection == null) {
                 try {
-                    initializeConnection(remoteAddress, authenticator);
-                } catch (Exception e) {
+                    connection = createSocketConnection(target);
+                } catch (IOException e) {
                     logger.finest(e);
-                } finally {
+                    callback.onFailure(e);
                     connectionsInProgress.remove(target);
+                    return;
                 }
-
             }
+
+            authenticate(target, connection, asOwner, callback);
+        }
+    }
+
+    private void authenticated(Address target, ClientConnection connection) {
+        if (connections.put(connection.getRemoteEndpoint(), connection) == null) {
+            fireConnectionAddedEvent(connection);
         }
+        connectionsInProgress.remove(target);
     }
+
+    private void failed(Address target, ClientConnection connection, Throwable throwable) {
+        logger.finest(throwable);
+        connection.close(throwable);
```

Comment:
What if connection is already there but authentication `asOwner` fails? Connection will be closed but is it going to be removed from connections map?